src/components/src/side-panel/layer-panel/layer-column-mode-config.tsx

```diff
@@ -0,0 +1,215 @@
+
+import React, {Fragment, useCallback, useMemo} from 'react';
+import styled from 'styled-components';
+
+import {
+  PanelHeaderContent,
+  PanelLabel,
+  StyledPanelHeader,
+  PanelContent
+} from '../../common/styled-components';
+import Checkbox from '../../common/checkbox';
+import {MinimalField} from '../../common/field-selector';
+import PanelHeaderActionFactory from '../panel-header-action';
+import LayerColumnConfigFactory from './layer-column-config';
+import {FormattedMessage} from '@kepler.gl/localization';
+import {assignColumnsByColumnMode, Layer, LayerInfoModal, LayerBaseConfig} from '@kepler.gl/layers';
+import {SupportedColumnModes, FieldPair} from '@kepler.gl/types';
+
+import {Help} from '../../common/icons';
+
+LayerColumnModeConfigFactory.deps = [LayerColumnConfigFactory, PanelHeaderActionFactory];
+
+const TopRow = styled.div`
+  display: flex;
+  justify-content: space-between;
+`;
+
+const Separator = styled(({children, className}) => (
+  <div className={className}>
+    <div className="separator-line-cell">
+      <div className="separator-line" />
+    </div>
+    <div className="separator-content">{children}</div>
+    <div className="separator-line-cell">
+      <div className="separator-line" />
+    </div>
+  </div>
+))`
+  display: flex;
+  flex-direction: row;
+  & > * + * {
+    margin-left: 10px;
+  }
+  & > .separator-line-cell {
+    display: flex;
+    flex-grow: 1;
+    align-items: center;
+  }
+  & > .separator-content {
+    color: ${props => props.theme.labelColor};
+  }
+  & > .separator-line-cell > .separator-line {
+    display: flex;
+    border-bottom: 1px solid ${props => props.theme.labelColor};
+    width: 100%;
+    line-height: 1px;
+    height: 1px;
+  }
+`;
+
+const ConfigPanesContainer = styled.div`
+  display: flex;
+  flex-direction: column;
+  & > * + * {
+    margin-top: 10px;
+  }
+  fieldset {
+    border: 1px solid ${props => props.theme.labelColor};
+    margin: 0;
+  }
+  legend {
+    display: flex;
+    label {
+      color: ${props => props.theme.labelColor};
+    }
+  }
+`;
+
+export type LayerColumnModeConfigProps<FieldOption extends MinimalField> = {
+  layer: Layer;
+  supportedColumnModes: SupportedColumnModes[] | null;
+  id: string;
+  layerConfig: LayerBaseConfig;
+  fields: FieldOption[];
+  fieldPairs?: FieldPair[];
+  openModal: (l: LayerInfoModal) => void;
+  updateLayerConfig: (config: Partial<LayerBaseConfig>) => void;
+};
+
+function LayerColumnModeConfigFactory(
+  LayerColumnConfig: ReturnType<typeof LayerColumnConfigFactory>,
+  PanelHeaderAction: ReturnType<typeof PanelHeaderActionFactory>
+) {
+  const LayerColumnModeConfig: React.FC<LayerColumnModeConfigProps<any>> = ({
+    layer,
+    supportedColumnModes,
+    id,
+    layerConfig,
+    fields,
+    fieldPairs,
+    openModal,
+    updateLayerConfig
+  }: LayerColumnModeConfigProps<any>) => {
+    const {columns} = layerConfig;
+    const columnModes = useMemo(
+      () =>
```

Comment:
Just a thought: That is a lot of processing in the UI. Could some of this logic could go into data container and similar utils? Or is it purely reformatting for display? Perhaps at least break out into helper functions?